### ISSUE TITLE
Squawk fails with “Recursion limit exceeded” on large or deeply nested SQL queries

### ISSUE THREAD

**Issue by Kirillov-da** [sbdchd/squawk#444]
Squawk fails to lint certain SQL migration files due to a recursion limit exceeded error. This appears when the SQL contains a very large or deeply nested query, such as a single-line INSERT INTO ... SELECT ... with multiple joins or subqueries.

**Steps to Reproduce:**
1. Create a .sql file with a large or complex SQL statement, such as:
INSERT INTO table (col1, col2, col3) 
SELECT a, b, c 
FROM ... 
JOIN ... 
WHERE ...;  -- All on a single line, with many nested clauses
2. Run Squawk against this file.
3. Observe the error: Schema error: recursion limit exceeded at line 1 column 9701

**Expected Behavior:**
Squawk should:
- Either successfully lint complex SQL files,
- Or fail gracefully with a warning rather than crashing the process.

**Actual Behavior:**
Squawk fails with a fatal parsing error and halts the pipeline or workflow using it.

**Environment:**
- Squawk version: 1.6.0 (built from source)
- Approximate size of failing SQL: single line > 9500 characters

**Error Log:**
Schema error: recursion limit exceeded at line 1 column 9701

**Impact:**
- Blocks CI/CD processes that rely on Squawk for SQL linting
- Forces developers to disable linting or artificially reformat valid queries

**Temporary Workaround:**
- Manually format SQL queries into multiple lines
- Skip linting for specific migration files or jobs

**Suggested Fix:**
- Increase internal recursion limit (if applicable), or
- Implement a more robust parser that can handle deeply nested queries, or
- Provide a configuration option to limit parser depth or skip large queries

**Comment by Kirillov-da** [comment]
@chdsbd I am unable to provide the full SQL statement at the moment due to confidentiality and size. However, I can share a similar query that causes the same error:
```
with tier_stat as (
    select 999001 as tournament_id ,1 as tier union all
    select 999002 as tournament_id ,2 as tier union all
    select 999003 as tournament_id ,2 as tier union all
    select 999004 as tournament_id ,1 as tier union all
    select 999005 as tournament_id ,2 as tier union all
    select 999006 as tournament_id ,2 as tier union all
    select 999007 as tournament_id ,1 as tier union all
    select 999008 as tournament_id ,1 as tier union all
    select 999009 as tournament_id ,2 as tier union all
    select 999010 as tournament_id ,1 as tier union all
    select 999011 as tournament_id ,2 as tier union all
    select 999012 as tournament_id ,1 as tier union all
    select 999013 as tournament_id ,2 as tier union all
    select 999014 as tournament_id ,2 as tier union all
    select 999015 as tournament_id ,1 as tier union all
    select 999016 as tournament_id ,1 as tier union all
    select 999017 as tournament_id ,2 as tier union all
    select 999018 as tournament_id ,1 as tier union all
    select 999019 as tournament_id ,2 as tier union all
    select 999020 as tournament_id ,1 as tier union all
    select 999021 as tournament_id ,2 as tier union all
    select 999022 as tournament_id ,2 as tier union all
    select 999023 as tournament_id ,1 as tier union all
    select 999024 as tournament_id ,2 as tier union all
    select 999025 as tournament_id ,1 as tier union all
    select 999026 as tournament_id ,1 as tier union all
    select 999027 as tournament_id ,2 as tier union all
    select 999028 as tournament_id ,1 as tier union all
    select 999029 as tournament_id ,2 as tier union all
    select 999030 as tournament_id ,1 as tier union all
    select 999031 as tournament_id ,2 as tier union all
    select 999032 as tournament_id ,2 as tier union all
    select 999033 as tournament_id ,1 as tier union all
    select 999034 as tournament_id ,2 as tier union all
    select 999035 as tournament_id ,1 as tier union all
    select 999036 as tournament_id ,2 as tier union all
    select 999037 as tournament_id ,1 as tier union all
    select 999038 as tournament_id ,2 as tier union all
    select 999039 as tournament_id ,1 as tier union all
    select 999040 as tournament_id ,2 as tier
)

UPDATE tree.tournaments t 
SET tier = ts.tier
FROM tier_stat AS ts
WHERE ts.tournament_id = t.tournament_id;
```
I hope this will help in identifying the issue. Let me know if you need further clarification or additional information!

**Comment by chdsbd** [comment]
@Kirillov-da that statement doesn't error in squawk 1.6.1.

**Comment by chdsbd** [comment]
@Kirillov-da I was able to modify your query and replicate the issue with this statement: https://gist.github.com/chdsbd/3c1ad3e4ce2e6ad7c6519c07077d4f36


```

  note: Squawk schema failed to parse Postgres response. This indicates a bug with Squawk. Please report this error to https://github.com/sbdchd/squawk. Schema error: recursion limit exceeded at line 1 column 9708
  help: Modify your Postgres statement to use valid syntax.

find detailed examples and solutions for each rule at https://squawkhq.com/docs/rules
Found 1 issue in 1 file (checked 1 source file)     
```

**Comment by Kirillov-da** [comment]
@chdsbd Yes, that's right, if you use a too nested SQL file, there will be an error, the essence of which I described above.

**Comment by chdsbd** [comment]
Instead of a CTE, you could write this statement as:


```sql
UPDATE tree.tournaments t 
SET tier = tournament_data.tier
FROM (
    VALUES 
    (999001, 1),
    (999002, 2),
    (999003, 2),
    (999004, 1),
    (999005, 2),
    (999006, 2),
    (999007, 1),
    (999008, 1),
    (999009, 2)
) AS tournament_data(tournament_id, tier)
WHERE t.tournament_id = tournament_data.tournament_id;

```

This kind of statement doesn't crash squawk

**Comment by chdsbd** [comment]
We're releasing a new major version of Squawk soon that includes a rebuilt parser written in Rust.

On a release build of Squawk v2, Squawk doesn't hit a stack overflow.

**Comment by chdsbd** [comment]
@Kirillov-da Can you try your SQL file with squawk 2.0.0?

**Comment by Kirillov-da** [comment]
@chdsbd In the nearest future we will test the tool in a real task and I will give you feedback for sure. Thank you for the work done and I hope that everything will work for us now!

**Comment by Kirillov-da** [comment]
Dear @chdsbd,
After upgrading to version 2.1.0 of Squawk, we encountered an issue with the absence of the --dump-ast flag, which was available in previous versions. When attempting to use this flag, the following error occurs:
```
error: Found argument '--dump-ast' which wasn't expected, or isn't valid in this context
USAGE:
    squawk [FLAGS] [OPTIONS] [path]... [SUBCOMMAND]
For more information try --help
```
We would like to inquire:
    1. Was the --dump-ast flag removed or changed in version 2.1.0?
    2. Is there a way to achieve similar functionality for working with AST in this version or upcoming releases?
    3. If possible, could you provide alternative methods for working with AST that are supported in version 2.1.0?
Thank you in advance for your help, and we look forward to your guidance on resolving this issue.

**Comment by Kirillov-da** [comment]
@chdsbd 
First of all, thank you for your continued work on Squawk — it’s a great tool and has been instrumental in maintaining SQL quality across our projects.
While upgrading from an earlier version of Squawk to a more recent release (v2.1.0+), we noticed that the --dump-ast flag appears to have been removed. This flag was critical for our CI pipeline as it allowed us to perform fine-grained analysis and filtering of SQL statements (e.g., rejecting disallowed operations such as UPDATE, INSERT, etc.).
We’ve looked through the changelogs and release notes but couldn’t find any mention of the deprecation or removal of this flag. We were wondering:
1) Was --dump-ast intentionally removed?
2) If so, is there an alternative mechanism that provides a similar AST-level output for downstream parsing/filtering?
3) Would it be possible to reintroduce this feature, even under a debug or advanced flag?
We’re currently evaluating whether we can adapt to --debug Parse as a partial replacement, but it doesn’t seem to offer the same level of structured output that --dump-ast provided.
Thanks again for your time and for the great tool. Looking forward to your thoughts.

P.S. Yes, we’ve tested our SQL file with Squawk 2.0.0. The good news is that the Schema error: recursion limit exceeded at line 1 column 9701 issue did not occur in this version.
However, while testing, we observed that the --dump-ast flag — which we rely on heavily for AST-level filtering of SQL statements — was already absent in 2.0.0, just as it is in 2.1.0 and later versions. This makes it challenging for us to maintain the same level of linting rigor we had previously.
Thanks again for your help and support!

**Comment by sbdchd** [comment]
Ah gotcha, yeah we removed the `--dump-ast` flag for now until the new parser & AST can replicate its functionality, we don't have much AST coverage at the moment.


Next steps are:

- setup ungrammar: https://github.com/rust-analyzer/ungrammar
- build out the description of the AST
- auto generate the AST
- expose the AST via the CLI


The `--debug=parse` option shows the CST (aka syntax tree) which is a layer below the AST. The CST isn't easy to work with and not intended for downstream consumers. The AST on the other hand is pretty straight forward.


I'm currently working on setting up ungrammar, but no ETA for when it will be added back to the CLI.


Do you know which specific AST nodes you're using for analysis? Maybe we could get something out in the meantime until we have full coverage of the postgresql grammar

**Comment by Kirillov-da** [comment]
@sbdchd
Thanks a lot for the detailed explanation — that makes sense now!
We primarily relied on --dump-ast to enforce a strict filter over top-level SQL statements in migration files. Specifically, we used it to allow only a limited set of statement types, and block potentially dangerous ones like UPDATE, DELETE, etc.
Here are the AST node types we were filtering for (or against):
     - UpdateStmt
     - InsertStmt
     - DeleteStmt
     - TruncateStmt
     - SelectStmt
If there's any way to expose even partial support for dumping these node types through the CLI (perhaps under an --experimental flag), it would help us maintain our linting checks without relying on the CST. 
Thanks for being open about the roadmap, and really appreciate all the effort you're putting into rebuilding the AST support — it's super helpful for folks like us relying on it!

**Comment by sbdchd** [comment]
@Kirillov-da curious how much detail you need in the AST output?

Specifically, if you had fields you're interested in

I've added ungrammar support, so now we have the initial scaffolding of the AST.

TODO:
- fill out the rest of the AST (it's only a skeleton right now for the most part)
- serialize the AST to JSON

**Comment by Kirillov-da** [comment]
@sbdchd 
Hey! Really appreciate the update — great to hear ungrammar is already integrated! 🙌

In our use case, we were relying on the --dump-ast output to detect disallowed top-level SQL statement types in migration files.

We were specifically filtering out:

    UpdateStmt

    InsertStmt

    DeleteStmt

    TruncateStmt

    sometimes SelectStmt

Here’s a simplified example of how we used the AST output:
```
squawk --dump-ast Parsed "$FILE" | jq -er '
  map(
    select(
      .stmt and type == "object" and 
      (.stmt | keys[] | test("^(UpdateStmt|InsertStmt|DeleteStmt|TruncateStmt|SelectStmt)$"))
    )
  ) | 
  if length > 0 then
    error("Disallowed operator used: " + ( .[] | .stmt | tostring ))
  else
    "Lint passed, only allowed statements are used"
  end
```

So essentially, we only needed the high-level type of each statement, like:
```
{
  "stmt": {
    "UpdateStmt": { ... }
  },
  "type": "object"
}
```
If we could get that structure (or even just the statement type name) back in a new AST dump, that would already fully cover our use case.

Let me know if you'd like us to test anything or provide more examples — happy to help!

Thanks again for pushing this forward 👏

**Comment by sbdchd** [comment]
Should be fixed in 2.9.0!

```shell
❯ npx squawk --version
squawk 2.9.0

❯ echo 'select' | npx squawk --debug=ast
{
  "stmts": [
    {
      "type": "SELECT"
    }
  ],
  "version": "0.1"
}
```

**Comment by Kirillov-da** [comment]
@sbdchd @chdsbd 
We just wanted to say a huge thank you for your work on the v2.10 release! 🙏
We're using Squawk in our CI pipeline to lint SQL migrations, and with earlier versions (like v2.9.0), we occasionally ran into issues — particularly jq: error (...) messages when trying to parse output for things like DROP TABLE. It seemed like the JSON output wasn't fully compatible with jq, which made debugging harder.
But with v2.10, everything works flawlessly:
✔️ Clean and informative error/warning output
✔️ Exact line numbers and clear rule identifiers
✔️ jq now parses the output perfectly
✔️ And the prebuilt binary works smoothly on Alpine (with a little help from gcompat)
We tested everything locally and in CI — and it’s working like a charm.
Squawk reliably catches unsafe operations (like DROP TABLE, missing IF NOT EXISTS, etc.), passes clean migrations, and even supports rule exclusions via /* squawk-exclude */. That's 🔥
Big kudos for the great documentation and helpful output formatting.
It makes it easy to introduce Squawk even in teams where not everyone is a SQL expert.
Thanks again for building such a great tool ❤️
